File: sdk/acx/ios/Sources/ACX/AdaptiveConsent.swift
import Foundation

public struct PurposeDefinition: Codable, Hashable {
    public let id: String
    public let category: String
    public let legalBasis: String
    public let defaultEnabled: Bool
    public let description: [String: String]
}

public struct LocaleTemplate: Codable {
    public let locale: String
    public let title: String
    public let summary: String
    public let bulletPoints: [String]
    public let acceptCta: String
    public let rejectCta: String
    public let manageCta: String
    public let footer: String
    public let variantOverrides: [String: LocaleOverride]?
}

public struct LocaleOverride: Codable {
    public let title: String?
    public let summary: String?
    public let bulletPoints: [String]?
    public let acceptCta: String?
    public let rejectCta: String?
    public let manageCta: String?
    public let footer: String?
}

public struct PolicyTemplatePack: Codable {
    public let policyId: String
    public let version: String
    public let defaultLocale: String
    public let purposes: [PurposeDefinition]
    public let locales: [String: LocaleTemplate]
}

public struct ConsentDialog: Codable {
    public let locale: String
    public let title: String
    public let summary: String
    public let bulletPoints: [String]
    public let acceptCta: String
    public let rejectCta: String
    public let manageCta: String
    public let footer: String
    public let purposes: [PurposeDefinition]
    public let policyId: String
    public let policyVersion: String
    public let variant: String
}

public struct PurposeScope: Codable {
    public let id: String
    public let enabled: Bool
}

public enum AdaptiveConsentError: Error {
    case darkPatternFound(pattern: String)
    case missingLocale
}

public final class AdaptiveConsent {
    private let pack: PolicyTemplatePack
    private let disallowedPatterns: [String]

    public init(pack: PolicyTemplatePack, disallowedPatterns: [String] = ["preselected", "you must accept"]) throws {
        self.pack = pack
        self.disallowedPatterns = disallowedPatterns
        try validate()
    }

    private func validate() throws {
        for (locale, template) in pack.locales {
            try inspect(locale: locale, template: template)
        }
    }

    private func inspect(locale: String, template: LocaleTemplate) throws {
        try inspect(locale: locale, value: template.title)
        try inspect(locale: locale, value: template.summary)
        try template.bulletPoints.forEach { try inspect(locale: locale, value: $0) }
        if let overrides = template.variantOverrides {
            for override in overrides.values {
                if let summary = override.summary {
                    try inspect(locale: locale, value: summary)
                }
            }
        }
    }

    private func inspect(locale: String, value: String) throws {
        for pattern in disallowedPatterns where value.lowercased().contains(pattern) {
            throw AdaptiveConsentError.darkPatternFound(pattern: pattern)
        }
    }

    public func render(locale: String, scopedPurposes: [String]? = nil) throws -> ConsentDialog {
        guard let template = pack.locales[locale] ?? pack.locales[pack.defaultLocale] else {
            throw AdaptiveConsentError.missingLocale
        }
        let purposes = scope(purposes: pack.purposes, requested: scopedPurposes)
        return ConsentDialog(
            locale: template.locale,
            title: template.title,
            summary: template.summary,
            bulletPoints: template.bulletPoints,
            acceptCta: template.acceptCta,
            rejectCta: template.rejectCta,
            manageCta: template.manageCta,
            footer: template.footer,
            purposes: purposes,
            policyId: pack.policyId,
            policyVersion: pack.version,
            variant: "control"
        )
    }

    private func scope(purposes: [PurposeDefinition], requested: [String]?) -> [PurposeDefinition] {
        guard let requested else { return purposes }
        let set = Set(requested)
        return purposes.filter { set.contains($0.id) }
    }
}

File: sdk/acx/android/src/main/kotlin/com/summit/acx/AdaptiveConsent.kt
package com.summit.acx

import java.security.KeyFactory
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

public data class PurposeDefinition(
    val id: String,
    val category: String,
    val legalBasis: String,
    val defaultEnabled: Boolean,
    val description: Map<String, String>
)

public data class LocaleOverride(
    val title: String? = null,
    val summary: String? = null,
    val bulletPoints: List<String>? = null,
    val acceptCta: String? = null,
    val rejectCta: String? = null,
    val manageCta: String? = null,
    val footer: String? = null
)

public data class LocaleTemplate(
    val locale: String,
    val title: String,
    val summary: String,
    val bulletPoints: List<String>,
    val acceptCta: String,
    val rejectCta: String,
    val manageCta: String,
    val footer: String,
    val variantOverrides: Map<String, LocaleOverride>? = null
)

public data class PolicyTemplatePack(
    val policyId: String,
    val version: String,
    val defaultLocale: String,
    val purposes: List<PurposeDefinition>,
    val locales: Map<String, LocaleTemplate>
)

public data class ConsentDialog(
    val locale: String,
    val title: String,
    val summary: String,
    val bulletPoints: List<String>,
    val acceptCta: String,
    val rejectCta: String,
    val manageCta: String,
    val footer: String,
    val purposes: List<PurposeDefinition>,
    val policyId: String,
    val policyVersion: String,
    val variant: String
)

public data class PurposeScope(val id: String, val enabled: Boolean)

public data class ConsentRecord(
    val policyId: String,
    val policyVersion: String,
    val userId: String,
    val locale: String,
    val decision: String,
    val purposes: List<PurposeScope>,
    val timestamp: String,
    val variant: String
)

public data class ConsentArtifact(
    val algorithm: String,
    val signature: String,
    val payload: ConsentRecord
)

public class DarkPatternException(message: String) : Exception(message)

public class AdaptiveConsent(
    private val pack: PolicyTemplatePack,
    private val disallowed: List<String> = listOf("preselected", "you must accept")
) {
    init {
        validate()
    }

    private fun validate() {
        pack.locales.forEach { (locale, template) ->
            inspect(locale, template.title)
            inspect(locale, template.summary)
            template.bulletPoints.forEach { inspect(locale, it) }
            template.variantOverrides?.values?.forEach { override ->
                override.summary?.let { inspect(locale, it) }
            }
        }
    }

    private fun inspect(locale: String, text: String) {
        disallowed.forEach {
            if (text.lowercase().contains(it)) {
                throw DarkPatternException("$locale contains disallowed pattern $it")
            }
        }
    }

    public fun render(locale: String, scopedPurposes: List<String>? = null): ConsentDialog {
        val template = pack.locales[locale] ?: pack.locales[pack.defaultLocale]
            ?: throw IllegalArgumentException("Locale $locale not found")
        val purposes = if (scopedPurposes == null) {
            pack.purposes
        } else {
            val set = scopedPurposes.toSet()
            pack.purposes.filter { set.contains(it.id) }
        }
        return ConsentDialog(
            locale = template.locale,
            title = template.title,
            summary = template.summary,
            bulletPoints = template.bulletPoints,
            acceptCta = template.acceptCta,
            rejectCta = template.rejectCta,
            manageCta = template.manageCta,
            footer = template.footer,
            purposes = purposes,
            policyId = pack.policyId,
            policyVersion = pack.version,
            variant = "control"
        )
    }

    public fun sign(record: ConsentRecord, privateKeyPem: String): ConsentArtifact {
        val privateKey = decodePrivateKey(privateKeyPem)
        val signature = Signature.getInstance("SHA256withRSA")
        signature.initSign(privateKey)
        signature.update(record.toString().toByteArray(Charsets.UTF_8))
        val signed = Base64.getEncoder().encodeToString(signature.sign())
        return ConsentArtifact("SHA256withRSA", signed, record)
    }

    public fun verify(artifact: ConsentArtifact, publicKeyPem: String): Boolean {
        val publicKey = decodePublicKey(publicKeyPem)
        val signature = Signature.getInstance("SHA256withRSA")
        signature.initVerify(publicKey)
        signature.update(artifact.payload.toString().toByteArray(Charsets.UTF_8))
        return signature.verify(Base64.getDecoder().decode(artifact.signature))
    }

    private fun decodePrivateKey(pem: String): PrivateKey {
        val content = pem
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replace("\n", "")
        val bytes = Base64.getDecoder().decode(content)
        val spec = PKCS8EncodedKeySpec(bytes)
        return KeyFactory.getInstance("RSA").generatePrivate(spec)
    }

    private fun decodePublicKey(pem: String): PublicKey {
        val content = pem
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
            .replace("\n", "")
        val bytes = Base64.getDecoder().decode(content)
        val spec = X509EncodedKeySpec(bytes)
        return KeyFactory.getInstance("RSA").generatePublic(spec)
    }
}
